### PR TITLE
sjawhar-legion-310: Replace Tailscale IPs with MagicDNS hostnames, switch NATS to host networking

### LINKED ISSUE #310: Replace hardcoded Tailscale IPs with MagicDNS hostnames in NATS and infra config
## Context

Part of #181 (Integrate Tailscale tsnet for listener auth). Independent of #308 and #309.

The Pulumi config (`packages/envoy/infra/Pulumi.prod.yaml`) contains hardcoded Tailscale IPs for each machine (e.g., `tailscaleIp: "100.117.175.119"`). These are used for:
1. NATS cluster routes (`nats://{ip}:6222`) — inter-node clustering
2. NATS client URLs (`nats://{ip}:4222`) — Go services connecting to remote NATS peers
3. The `MachineConfig.tailscaleIp` field in Pulumi infrastructure code

The IPs were "rendered per-host but accidentally committed" (from #181). They should be replaced with MagicDNS hostnames for stability (IPs can change, hostnames are stable) and to avoid committing network addresses.

**Key invariant:** `machine.name` in the Pulumi config IS the canonical MagicDNS hostname for each machine. The existing `sshHost` values (e.g., `ssh://sami@sami`, `ssh://claude@sami-claude`) already use these names, confirming MagicDNS resolution works at the host level. If any machine's MagicDNS hostname does not match its `machine.name`, do **not** overload `name` — add a dedicated `hostname` field to MachineConfig instead.

**Note:** `packages/envoy/docs/constraints.md` currently says "Use Tailscale IPs for inter-machine traffic" AND "Keep ports and public ingress limited to the already-open webhook receiver ports." Both lines need revision since this issue changes inter-machine addressing and NATS host networking exposes ports directly on the host interface.

**Learning reference:** `docs/solutions/envoy/docker-tailscale-networking.md` documents that Docker bridge networking cannot resolve MagicDNS hostnames. This issue sidesteps that by switching NATS to host networking. The learning must be updated to note the host networking approach.

This sub-issue can be done independently of the tsnet work (#308, #309).

### NATS containers: switch to host networking

Currently NATS containers use Docker bridge networking with port bindings. Bridge networking can't resolve MagicDNS hostnames (documented in `docs/solutions/envoy/docker-tailscale-networking.md`). Switching to host networking lets NATS resolve MagicDNS peers directly.

**Changes to `packages/envoy/infra/nats.ts`:**
- Remove explicit `ports` array (not needed with host networking)
- Add `networkMode: "host"` to NATS container
- In `computeNatsRoutes()`: use `m.name` (MagicDNS hostname) instead of `m.tailscaleIp`
- Remove `tailscaleIp` from the local `PeerInfo` interface

**Changes to `packages/envoy/infra/services.ts`:**
- In `computeNatsUrls()`: use `m.name` (MagicDNS hostname) instead of `m.tailscaleIp`
- Remove `tailscaleIp` from the local `PeerInfo` interface
- Update `getNatsUrls()` helper to stop mapping `tailscaleIp`

### Remove `tailscaleIp` from MachineConfig

- Remove the `tailscaleIp` field from `MachineConfig` interface in `packages/envoy/infra/machines.ts`
- Remove `tailscaleIp` entries from all 4 machines in `packages/envoy/infra/Pulumi.prod.yaml`

### Update deploy compose files

- Update `packages/envoy/deploy/compose/nats/peer.compose.yml`: switch to `network_mode: host`, remove `ports` section
- Update `packages/envoy/deploy/compose/nats/compose.yml`: same changes (host networking, remove ports)
- `packages/envoy/deploy/scripts/render-nats-peer.sh` is already format-agnostic — it takes `NATS_ROUTES` from env vars. No code changes needed; the caller just needs to pass hostname-based routes instead of IP-based routes.
- Verify `packages/envoy/deploy/scripts/up-nats-peer.sh` works with hostname-based routes (it shells out to `render-nats-peer.sh` + `docker compose up`)

### Update unit tests

- Update `packages/envoy/infra/__tests__/nats.test.ts`: remove `tailscaleIp` from `PeerInfo` test fixtures, update expected routes to use machine names instead of IPs
- Update `packages/envoy/infra/__tests__/services.test.ts`: same — remove `tailscaleIp` from fixtures, update expected URLs to use machine names for remote peers

### Update documentation

- **`packages/envoy/infra/README.md`**: Update Machine Configuration table — remove `tailscaleIp` row, note that `machine.name` must match the MagicDNS hostname
- **`packages/envoy/docs/constraints.md`**: Update TWO lines:
  1. Change "Use Tailscale IPs for inter-machine traffic" → "Use MagicDNS hostnames for inter-machine traffic (hostnames are stable; IPs can change)"
  2. Update "Keep ports and public ingress limited to the already-open webhook receiver ports" to acknowledge that NATS now uses host networking (ports 4222, 6222, 8222 are exposed on host interfaces, protected by Tailscale ACLs / firewall rules)
- **`docs/solutions/envoy/docker-tailscale-networking.md`**: Add a note that NATS now uses host networking (sidesteps the bridge DNS limitation). The core rule about bridge networking still applies to any future bridge-networked containers, but NATS specifically is no longer affected.

## Acceptance Criteria

- [ ] No hardcoded Tailscale IPs in active runtime code, config, scripts, or current documentation (historical docs/learnings may reference IPs in examples)
- [ ] `MachineConfig` no longer has a `tailscaleIp` field
- [ ] `PeerInfo` interfaces in `nats.ts` and `services.ts` no longer have `tailscaleIp`
- [ ] NATS containers use host networking (`networkMode: "host"` in Pulumi, `network_mode: host` in compose)
- [ ] NATS Pulumi container has no explicit `ports` array (host networking makes it unnecessary)
- [ ] NATS cluster routes use MagicDNS hostnames (e.g., `nats://sami-agents-mx:6222`)
- [ ] Go service NATS client URLs use MagicDNS hostnames for remote peers
- [ ] Local NATS peer still uses `127.0.0.1` for same-machine connections
- [ ] Pulumi infra tests pass with updated config (`computeNatsRoutes`, `computeNatsUrls`, `renderNatsConf`)
- [ ] `packages/envoy/infra/README.md` updated (remove `tailscaleIp`, document `machine.name` = MagicDNS hostname)
- [ ] `packages/envoy/docs/constraints.md` updated — both the Tailscale IP rule and the port/ingress rule
- [ ] `docs/solutions/envoy/docker-tailscale-networking.md` updated with host networking note
- [ ] JetStream data survives the NATS networking change (named volumes persist)
- [ ] NATS cluster forms correctly with hostname-based routes

### Verification Commands (for runtime criteria)

```bash
# Verify NATS cluster formation (on each peer):
# Expected: num_routes = (total_nats_peers - 1), route URLs contain hostnames not IPs
curl -s http://127.0.0.1:8222/routez | jq '{num_routes: .num_routes, routes: [.routes[].url]}'

# Verify JetStream health (on each peer):
# Compare stream names and message counts before and after migration
curl -s http://127.0.0.1:8222/jsz | jq '{streams: .streams, messages: .messages}'

# Verify rendered NATS config uses hostnames:
docker exec envoy-nats cat /etc/nats/nats.conf | grep -E 'nats://'
```

### Rollout Criteria

Before deploying, verify:
- [ ] MagicDNS hostnames match `machine.name` values: run `nslookup sami-agents-mx`, `nslookup sami`, `nslookup sami-claude`, `nslookup ghost-wispr` from each host to confirm resolution
- [ ] Host ports 4222, 6222, 8222 are free on all machines: `ss -tlnp | grep -E ':(4222|6222|8222)'` (should show only the existing NATS container or nothing)
- [ ] Firewall/ACL posture verified: confirm UFW rules or Tailscale ACLs allow inter-peer traffic on ports 4222, 6222, 8222 between the 4 machines
- [ ] Capture pre-migration JetStream state: `curl -s http://127.0.0.1:8222/jsz | jq .` on each peer
- [ ] NATS cluster reforms with full JetStream replication after migration: `/routez` shows expected route count, `/jsz` shows same streams and messages as pre-migration

### Testing Infrastructure

**Available:**
- Pulumi infra tests (`packages/envoy/infra/__tests__/`) — cover `computeNatsRoutes()`, `computeNatsUrls()`, and `renderNatsConf()`
- NATS monitoring endpoint (`/routez`, `/jsz`) for verifying cluster health

**Gaps:**
- MagicDNS resolution from host-networked containers requires a Tailscale-connected host — can't test in CI
- NATS cluster formation test requires multiple machines — manual verification during rollout
- JetStream data persistence across networking change — manual verification using pre/post `/jsz` comparison

## Implementation Notes

- The NATS health check uses `wget` to `http://127.0.0.1:8222/healthz`. With host networking this still works.
- The listener already uses host networking (`networkMode: "host"` in services.ts), so this makes NATS consistent with the rest of the fleet.
- The first machine (`sami-agents-mx`) has no `sshHost` because Pulumi runs locally on it. Its MagicDNS name should still match `machine.name`.
- `render-nats-peer.sh` is already format-agnostic — it takes `NATS_ROUTES` from environment and writes them to `nats.conf`. No code changes needed in the script itself.
- Host networking does not create new port exposure (the same ports were already published via Docker port mappings), but it does remove container network namespace isolation. The ports are protected by Tailscale's network boundary and any host-level firewall rules.
- **Escalation trigger:** If any machine's MagicDNS hostname does not match its `machine.name`, do NOT overload the name field. Instead, replace `tailscaleIp` with an explicit `hostname` field on MachineConfig.

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #310

## Summary

Replace hardcoded Tailscale IPs with stable MagicDNS hostnames for NATS cluster routing and client URLs, and switch NATS containers from Docker bridge networking to host networking.

### Changes
- **`MachineConfig`**: Removed `tailscaleIp` field from interface and all 4 machine entries in `Pulumi.prod.yaml`
- **NATS (`nats.ts`)**: Removed `ports` array, added `networkMode: "host"`, routes now use `m.name` instead of `m.tailscaleIp`
- **Services (`services.ts`)**: NATS client URLs now use `m.name` for remote peers (local peer still `127.0.0.1`)
- **Compose files**: Both `compose.yml` and `peer.compose.yml` switched from `ports` to `network_mode: host`
- **Tests**: Updated fixtures and expectations to use hostnames instead of IPs
- **Docs**: Updated `constraints.md`, `README.md`, and `docker-tailscale-networking.md` to reflect the new addressing model

### Rollout notes
Before deploying, verify:
- MagicDNS hostnames match `machine.name` values (`nslookup` from each host)
- Host ports 4222, 6222, 8222 are free on all machines
- Capture pre-migration JetStream state via `/jsz`
- After migration, verify cluster forms via `/routez` and JetStream replication via `/jsz`

### CODE REVIEW

**Comment by sjawhar** [comment]
## Behavioral Test Results

### Summary
**PASS** — 12/12 CI-verifiable acceptance criteria verified. 2 additional criteria (13, 14) are manual rollout verification only, as documented in the issue.

### Results
| # | Criterion | Status | Evidence |
|---|-----------|--------|----------|
| 1 | No hardcoded Tailscale IPs in active runtime code/config/docs | ✅ | `grep -rn "tailscaleIp" packages/envoy/infra/ packages/envoy/infra/__tests__/` → no matches. `grep -rn "100\.117\.175\|100\.100\.92\|100\.85\.18\|100\.103\.243" packages/envoy/` → no matches. Only remaining IP example is in `docs/solutions/envoy/docker-tailscale-networking.md:75` (historical learning, allowed by spec). |
| 2 | `MachineConfig` no `tailscaleIp` field | ✅ | `machines.ts:17-24` — interface has `name`, `sshHost?`, `machineId`, `nats?`, `listener`, `receivers?`. No `tailscaleIp`. |
| 3 | `PeerInfo` interfaces no `tailscaleIp` | ✅ | `nats.ts:6-9`: `{name: string, nats: boolean}`. `services.ts:7-10`: `{name: string, nats: boolean}`. No `tailscaleIp` in either. |
| 4 | NATS containers use host networking | ✅ | Pulumi: `nats.ts:89` → `networkMode: "host"`. Compose: `compose.yml:5` → `network_mode: host`. `peer.compose.yml:5` → `network_mode: host`. Verified via `docker compose config` — both show `network_mode: host`, no `ports` block. |
| 5 | NATS Pulumi container has no `ports` array | ✅ | `nats.ts:83-114` — container definition has `networkMode: "host"`, `uploads`, `volumes`, `command`, `healthcheck`. No `ports` property present. |
| 6 | NATS cluster routes use MagicDNS hostnames | ✅ | `nats.ts:18` → `` `nats://${m.name}:6222` ``. Test: `nats.test.ts:14` → `expect(routes).toEqual(["nats://sami:6222", "nats://sami-claude:6222"])`. |
| 7 | Go service NATS client URLs use MagicDNS hostnames | ✅ | `services.ts:24` → `` `nats://${m.name}:4222` `` for remote peers. `getNatsUrls()` at `services.ts:40-46` maps machines and feeds to `computeNatsUrls()`. `NATS_URLS` injected into listener (`:81`), github (`:132`), slack (`:175`), ghostwispr (`:209`) containers. Test: `services.test.ts:19` → `"nats://sami-agents-mx:4222,nats://sami:4222,nats://sami-claude:4222"`. |
| 8 | Local NATS peer uses `127.0.0.1` | ✅ | `services.ts:27` → `["nats://127.0.0.1:4222", ...remotePeers].join(",")`. Test: `services.test.ts:14` → `"nats://127.0.0.1:4222,nats://sami:4222,nats://sami-claude:4222"`. |
| 9 | Pulumi infra tests pass | ✅ | `bun test packages/envoy/infra/__tests__/` → **7 pass, 0 fail, 16 expect() calls** across 2 files (648ms). |
| 10 | README.md updated | ✅ | `README.md:104` → `name` documented as "Machine identifier — must match the machine's Tailscale MagicDNS hostname (used for NATS cluster routing)". No `tailscaleIp` row in table. |
| 11 | constraints.md updated (both lines) | ✅ | Line 21: "Use MagicDNS hostnames for inter-machine traffic (hostnames are stable; IPs can change)". Line 24: "Keep ports and public ingress limited to the already-open webhook receiver ports. NATS uses host networking (ports 4222, 6222, 8222 exposed on host interfaces, protected by Tailscale ACLs / firewall rules)." |
| 12 | docker-tailscale-networking.md updated | ✅ | Lines 84-94: "Update: NATS Host Networking (April 2026)" section added. Pattern section (48-51) updated with `name` field. Bullet points (54-57) updated for host-networked containers. Rule (61-63) notes host-networked containers since April 2026. |
| 13 | JetStream data survives networking change | ⏭️ | Manual rollout verification only (per issue spec). **Code supports it**: named volume `nats_nats_data` preserved in Pulumi (`nats.ts:78`), `nats_data` volume in `peer.compose.yml:12`. Volume name unchanged from pre-migration. |
| 14 | NATS cluster forms with hostname-based routes | ⏭️ | Manual rollout verification only (per issue spec). **Code supports it**: `computeNatsRoutes()` generates hostname-based routes, `renderNatsConf()` writes them to config, tests verify route format. |

### Additional Checks
| Check | Result |
|-------|--------|
| Type check (`bunx tsc --noEmit`) | ✅ Clean (only pre-existing TS5107 `moduleResolution=node10` deprecation) |
| Lint (`bunx biome check`) | ✅ 9 files checked, no issues |
| Compose validation (`docker compose config`) | ✅ Both files render with `network_mode: host`, no `ports` |

### Test Quality Critique
- **Coverage scope**: Tests cover `computeNatsRoutes`, `computeNatsUrls`, and `renderNatsConf` — the three pure functions that changed. 7 tests, 16 assertions.
- **No excessive mocking**: Tests exercise pure functions directly with real-ish machine names matching production values (sami-agents-mx, sami, sami-claude, ghost-wispr). No mocks at all. ✅
- **No circular logic**: Tests assert known expected output from known input, not re-implementing production code. ✅
- **P2: Happy-path-only coverage**: No edge case tests for empty machines array, all machines with `nats: false`, single-machine fleet, or machine names with special characters. Tests verify the common cases well but don't probe boundaries.
- **P2: No integration test for Pulumi wiring**: `createNatsPeer()` maps `MachineConfig` → `PeerInfo` → routes → conf → container, but no test verifies this full pipeline. Only the individual pure functions are tested.
- **No hardcoded expected values copied from implementation**: Expected values derive from the machine fixture data, not from running the code and capturing output. ✅

### Documentation Feedback
- README Machine Configuration table is clear — the `name` → MagicDNS contract is explicit on line 104. ✅
- constraints.md updates both required lines cleanly. ✅
- docker-tailscale-networking.md preserves the bridge networking rule for future containers while noting the NATS exception. Good structure. ✅
- No gaps or confusion found. Setup instructions (`bun install`, `pulumi up`) are accurate and referenced in Quick Start section.

### Observations
- **Deviation**: Implementer added `receivers.ghostwispr` row to README table (line 111). Not in spec but harmless — documenting an existing field.
- Pre-existing `moduleResolution=node10` deprecation warning in infra tsconfig (TS5107). Unrelated to this PR.
- `deleteBeforeReplace: true` correctly set on NATS container (`nats.ts:113`) — mandatory for host-networked containers per `docs/solutions/envoy/pulumi-remote-docker-patterns.md`.

### Config Compliance
- require_specific_task: false — not configured
- require_taiga_evidence: false — not configured

File: packages/envoy/infra/nats.ts [hunk 1]
```diff
@@ -5,7 +5,6 @@ import type { MachineConfig } from "./machines";
 
 interface PeerInfo {
   name: string;
-  tailscaleIp: string;
   nats: boolean;
```
Comment (sjawhar) [inline packages/envoy/infra/nats.ts]: **[P3 — MINOR]:** No integration test covers the full `createNatsPeer()` pipeline (MachineConfig → PeerInfo → routes → conf → container). Only the pure functions are unit tested. Pre-existing gap, not introduced by this PR.
